core/server/worker/src/main/java/alluxio/worker/grpc/BlockReadHandler.java

```diff
@@ -19,72 +19,469 @@
 import alluxio.retry.TimeoutRetry;
 import alluxio.security.authentication.AuthenticatedUserInfo;
+import alluxio.util.LogUtils;
 import alluxio.util.logging.SamplingLogger;
 import alluxio.worker.block.BlockWorker;
 import alluxio.worker.block.UnderFileSystemBlockReader;
 import alluxio.worker.block.io.BlockReader;
 
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Meter;
 import com.google.common.base.Preconditions;
+import com.google.protobuf.UnsafeByteOperations;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.internal.SerializingExecutor;
+import io.grpc.stub.CallStreamObserver;
 import io.grpc.stub.StreamObserver;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.PooledByteBufAllocator;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.nio.channels.FileChannel;
+import java.util.concurrent.Executor;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.locks.ReentrantLock;
 
 import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
 import javax.annotation.concurrent.NotThreadSafe;
 
 /**
- * This handler handles block read request. Check more information in {@link AbstractReadHandler}.
+ * This class handles {@link ReadRequest}s.
+ *
+ * Protocol: Check {@link alluxio.client.block.stream.GrpcDataReader} for additional information.
+ * 1. Once a read request is received, the handler creates a {@link DataReader} which reads
+ *    chunks of data from the block worker and pushes them to the buffer.
+ * 2. The {@link DataReader} pauses if there are too many packets in flight, and resumes if there
+ *    is room available.
+ * 3. The channel is closed if there is any exception during the data read/write.
+ *
+ * Threading model:
+ * Only two threads are involved at a given point of time: gRPC event thread, data reader thread.
+ * 1. The gRPC event thread accepts the read request, handles write callbacks. If any exception
+ *    occurs (e.g. failed to read from stream or respond to stream) or the read request is cancelled
+ *    by the client, the gRPC event thread notifies the data reader thread.
+ * 2. The data reader thread keeps reading from the file and writes to buffer. Before reading a
+ *    new data chunk, it checks whether there are notifications (e.g. cancel, error), if
+ *    there is, handle them properly. See more information about the notifications in the javadoc
+ *    of {@link ReadRequestContext} about CANCEL, EOF, and ERROR flags.
+ *
+ * @see ReadRequestContext
  */
-@edu.umd.cs.findbugs.annotations.SuppressFBWarnings(value = "BC_UNCONFIRMED_CAST_OF_RETURN_VALUE",
-    justification = "false positive with superclass generics, "
-        + "see more description in https://sourceforge.net/p/findbugs/bugs/1242/")
 @NotThreadSafe
-public final class BlockReadHandler extends AbstractReadHandler<BlockReadRequestContext> {
```

Comment:
Combine `BlockReadHandler` and `AbstractReadHandler`. There is no other class extending `AbstractReadHandler`